File: app/src/main/java/com/alexkn/syntact/domain/repository/BucketRepository.java
package com.alexkn.syntact.domain.repository;

import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.work.Data;
import androidx.work.ExistingWorkPolicy;
import androidx.work.OneTimeWorkRequest;
import androidx.work.WorkManager;

import com.alexkn.syntact.app.Property;
import com.alexkn.syntact.data.common.AppDatabase;
import com.alexkn.syntact.data.dao.BucketDao;
import com.alexkn.syntact.data.model.Bucket;
import com.alexkn.syntact.data.model.views.BucketDetail;
import com.alexkn.syntact.domain.worker.CreateBucketWorker;
import com.alexkn.syntact.restservice.SyntactService;
import com.alexkn.syntact.restservice.Template;

import java.time.Instant;
import java.util.Arrays;
import java.util.List;
import java.util.Locale;
import java.util.stream.Collectors;

import javax.inject.Inject;
import javax.inject.Singleton;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

@Singleton
public class BucketRepository {

    private static final String TAG = BucketRepository.class.getSimpleName();

    private final BucketDao bucketDao;

    private final SyntactService syntactService;

    private Property property;

    @Inject
    BucketRepository(SyntactService syntactService, Property property, Context context) {

        this.syntactService = syntactService;
        this.property = property;
        bucketDao = AppDatabase.Companion.getDatabase(context).bucketDao();
    }

    public List<Locale> getAvailableLanguages() {

        String[] languages = property.get("available-languages").split(",");
        return Arrays.stream(languages).map(Locale::new).collect(Collectors.toList());
    }

    public LiveData<List<Template>> findAvailableTemplates() {

        MutableLiveData<List<Template>> templates = new MutableLiveData<>();
        String token = "Token " + property.get("api-auth-token");
        syntactService.getTemplates(token).enqueue(new Callback<List<Template>>() {

            @Override
            public void onResponse(@NonNull Call<List<Template>> call, @NonNull Response<List<Template>> response) {

                if (response.isSuccessful()) {
                    Log.i(TAG, "Found " + response.body().size() + " templates");
                    templates.postValue(response.body());
                }
            }

            @Override
            public void onFailure(@NonNull Call<List<Template>> call, @NonNull Throwable t) {

            }
        });
        return templates;
    }

    public void addBucketWithNewTemplate(String name, Locale language, List<String> words) {

        Data data = new Data.Builder().putString("name", name).putString("language", language.getLanguage())
                .putStringArray("words", words.toArray(new String[0])).build();

        OneTimeWorkRequest workRequest = new OneTimeWorkRequest.Builder(CreateBucketWorker.class).setInputData(data).build();
        WorkManager.getInstance().enqueueUniqueWork(CreateBucketWorker.class.getName(), ExistingWorkPolicy.KEEP, workRequest);
    }

    public void addBucketWithExistingTemplate(Locale language, Template template) {

        Locale sourceLanguage = Locale.getDefault();

        Bucket bucket = new Bucket();
        bucket.setCreatedAt(Instant.now());
        bucket.setLanguage(language);
        bucket.setUserLanguage(sourceLanguage);
        bucket.setPhrasesUrl(template.getPhrasesUrl());
        bucket.setItemCount(template.getCount());
        bucketDao.insert(bucket);
    }

    public void removeLanguage(Bucket bucket) {

        bucketDao.delete(bucket);
    }

    public LiveData<Bucket> getBucket(Long id) {

        return bucketDao.findBucket(id);
    }

    public LiveData<List<BucketDetail>> getBucketDetails() {

        return bucketDao.findBucketDetails();
    }
}

File: app/src/main/java/com/alexkn/syntact/domain/repository/BucketRepository.kt
package com.alexkn.syntact.domain.repository

import androidx.lifecycle.LiveData
import androidx.work.*
import com.alexkn.syntact.app.Property
import com.alexkn.syntact.data.dao.BucketDao
import com.alexkn.syntact.data.dao.PlayerStatsDao
import com.alexkn.syntact.data.dao.SolvableItemDao
import com.alexkn.syntact.data.model.Bucket
import com.alexkn.syntact.data.model.SolvableItem
import com.alexkn.syntact.data.model.Template
import com.alexkn.syntact.data.model.views.BucketDetail
import com.alexkn.syntact.data.model.views.PlayerStats
import com.alexkn.syntact.domain.worker.CreateBucketWorker
import com.alexkn.syntact.rest.service.SyntactService
import com.alexkn.syntact.rest.to.TemplateResponse
import com.google.common.util.concurrent.ListenableFuture
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.util.*
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class BucketRepository @Inject
internal constructor(
        private val syntactService: SyntactService,
        private val property: Property,
        private val bucketDao: BucketDao,
        private val playerStatsDao: PlayerStatsDao,
        private val solvableItemDao: SolvableItemDao,
        private val solvableItemRepository: SolvableItemRepository
) {

    val availableLanguages: MutableList<Locale> = property["available-languages"].split(",").map { Locale(it) }.toMutableList()


    val bucketDetails: LiveData<List<BucketDetail>>
        get() = bucketDao.findBucketDetails()

    suspend fun findAvailableTemplates(): List<TemplateResponse> {

        val token = "Token " + property["api-auth-token"]
        return syntactService.getTemplates(token)
    }

    fun addBucketWithNewTemplate(name: String, language: Locale, words: List<String>): ListenableFuture<Operation.State.SUCCESS> {

        val data = Data.Builder().putString("name", name).putString("language", language.language)
                .putStringArray("words", words.toTypedArray()).build()

        val workRequest = OneTimeWorkRequest.Builder(CreateBucketWorker::class.java).setInputData(data).build()
        return WorkManager.getInstance().enqueueUniqueWork(CreateBucketWorker::class.java.name, ExistingWorkPolicy.KEEP, workRequest).result
    }

    suspend fun addBucketWithExistingTemplate(template: Template)  = withContext(Dispatchers.Default) {

        val sourceLanguage = Locale.getDefault()

        val bucket = Bucket(
                id = template.id,
                name = template.name,
                language = template.language,
                userLanguage = sourceLanguage,
                phrasesUrl = template.phrasesUrl,
                itemCount = template.count
        )
        val id = bucketDao.insert(bucket)
        val token = "Token " + property["api-auth-token"]

        val phrases = syntactService.getPhrases(token, bucket.phrasesUrl)

        val solvableItems = phrases.map {
            SolvableItem(
                    id = it.id,
                    text = it.text.capitalize(),
                    bucketId = id,
                    translationUrl = it.translationsUrl
            )
        }

        solvableItemDao.insert(solvableItems)
    }

    suspend fun removeLanguage(bucket: Bucket) {

        bucketDao.delete(bucket)
    }

    suspend fun deleteBucket(id: Long) {
        bucketDao.delete(id)
    }

    fun getBucket(id: Long): LiveData<Bucket> {

        return bucketDao.findBucket(id)
    }

    fun getBucketDetail(id: Long): LiveData<BucketDetail> {
        return bucketDao.findBucketDetail(id)
    }

    fun getPlayerStats(): LiveData<PlayerStats> {
        return playerStatsDao.find()
    }
}
